File: Palindrome_Reorder.py
class test:
    def __init__(self,n):
        print(n)
s = input()

# letters=set(s)
lettercount={x:s.count(x) for x in s}
sorted(lettercount.items(), key=lambda item: item[1])
# print(lettercount)

oddCount=0

oddChar=''
for x in lettercount.items():
    if(x[1]%2 == 0):
        continue
    oddCount+=1
    oddChar+=x[0]
    
if(oddCount>1):
    print("NO SOLUTION")
    exit()
ans=""
for x in lettercount.items():
    buffer=""
    for i in range(0,x[1]):
        buffer+=x[0]
    ans+=buffer[0:int(len(buffer)/2)]

if(len(s)%2 == 0):
    print(ans+ans[::-1])
else:
    print(ans+oddChar+ans[::-1])

File: Palindrome_Reorder.cpp
#include<iostream>
#include<bits/stdc++.h>
#define ll long long
#define rep(i,n) for(int i=0;i<n;i++)

using namespace std;

void solve(){
   //develop cpp algorithm from python solution
   string s;
   cin>>s;

   ll arr[26]{0};

   for(int i=0;i<s.length();i++){
        // cout<<s[i];
        // cout<<int(s[i]);
        arr[int(s[i])-65]++;
   }
//    int n=26;
//    rep(i,n)cout<<arr[i];
    ll oddcount=0;
    string ans;
    char oddChar;
   for(int i=0;i<26;i++){
        ll count =arr[i];
        ll printCount = count/2;
        if(arr[i]==0){
            continue;
        }
        if(arr[i]%2!=0){
            oddcount++;
            oddChar+=char(i+65);
        }
        if(oddcount>1){
            cout<<"NO SOLUTION";
            return;
        }
        
        for(int j=0;j<printCount;j++){
            ans+=char(i+65);
        }
   }
   
   //print the reverse string
   string temp = ans;
   reverse(ans.begin(),ans.end());
    if(oddcount){
        cout<<temp+oddChar+ans;
   }
   else{
    cout<<temp+ans;
   }
   
}

int main(int argc, char const *argv[])
{
    solve();
    return 0;
}
